File: mergeSorted.java

// Time Complexity : O(n+m)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
//used two pointer to traverse in descending order and 
// fill all the values in the decreaseing order

// Your code here along with comments explaining your approach

class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int n1 = m - 1;
        int n2 = n - 1;
        while (true) {
            if (n2 == -1)
                return;
            if (n1 == -1)
                break;
            if (nums2[n2] > nums1[n1])
                nums1[n1 + n2 + 1] = nums2[n2--];
            else
                nums1[n1 + n2 + 1] = nums1[n1--];

        }
        while (n2 != -1) {
            nums1[n2] = nums2[n2];
            n2--;
        }
        return;
    }
}

File: mergeSorted.py
# https://leetcode.com/problems/merge-sorted-array/
# Time O(m+n) Total length of nums1
# Space O(1) : No extra space used, modified in-place

class Solution:
    def merge(self, nums1 , m, nums2, n):
        """
        Do not return anything, modify nums1 in-place instead.
        """
        if not m:
            for i in range(n):
                nums1[i] = nums2[i]
        if not n:
            return
        
        p1, p2, last = m-1, n-1, len(nums1) - 1
        
        while p2 >= 0 and p1 >= 0:
            if nums2[p2] >= nums1[p1]:
                nums1[last] = nums2[p2]
                p2 -= 1
                
            else:
                nums1[last] = nums1[p1]
                p1 -= 1
            last -= 1
        
        while p2 >= 0:
            nums1[last] = nums2[p2]
            last -= 1
            p2 -= 1

nums1 = [1,2,3,0,0,0]
m = 3
nums2 = [2,5,6]
n = 3
s = Solution()
output = s.merge(nums1, m, nums2, n)
print(nums1)